internal/backends/mcu/pico_st7789.rs

```diff
@@ -124,29 +132,148 @@ pub fn init() {
+}
+
+impl<TO: WriteTarget<TransmittedWord = u8> + FullDuplex<u8>, CH: SingleChannel>
+    PioTransfer<TO, CH>
+{
+    fn wait(self) -> (CH, &'static mut [super::TargetPixel], TO) {
+        match self {
+            PioTransfer::Idle(a, b, c) => (a, b, c),
+            PioTransfer::Running(dma) => {
+                let (a, b, mut to) = dma.wait();
+                // After the DMA operated, we need to empty the receive FIFO, otherwise the touch screen
+                // driver will pick wrong values. Continue to read as long as we don't get a Err(WouldBlock)
+                while !to.read().is_err() {}
+                (a, b.0, to)
+            }
+        }
+    }
 }
 
-struct PicoDevices<Display, Touch> {
+struct PicoDevices<Display, Touch, PioTransfer, Stolen> {
     display: Display,
     touch: Touch,
     last_touch: Option<i_slint_core::graphics::Point>,
     timer: Timer,
+    buffer: &'static mut [super::TargetPixel],
+    pio: Option<PioTransfer>,
+    stolen_pin: Stolen,
 }
 
-impl<Display: Devices, IRQ: InputPin, CS: OutputPin<Error = IRQ::Error>, SPI: Transfer<u8>> Devices
-    for PicoDevices<Display, xpt2046::XPT2046<IRQ, CS, SPI>>
+impl<
+        DI: display_interface::WriteOnlyDataCommand,
+        RST: OutputPin<Error = Infallible>,
+        IRQ: InputPin<Error = Infallible>,
+        CS: OutputPin<Error = Infallible>,
+        SPI: Transfer<u8>,
+        TO: WriteTarget<TransmittedWord = u8> + FullDuplex<u8>,
+        CH: SingleChannel,
+        DC_: OutputPin<Error = Infallible>,
+        CS_: OutputPin<Error = Infallible>,
+    > Devices
+    for PicoDevices<
+        st7789::ST7789<DI, RST>,
+        xpt2046::XPT2046<IRQ, CS, SPI>,
+        PioTransfer<TO, CH>,
+        (DC_, CS_),
+    >
 {
     fn screen_size(&self) -> PhysicalSize {
-        self.display.screen_size()
+        let s = self.display.size();
+        euclid::size2(s.width as _, s.height as _)
+    }
+
+    fn render_line(
+        &mut self,
+        line: PhysicalLength,
+        dirty_region: crate::renderer::DirtyRegion,
+        fill_buffer: &mut dyn FnMut(&mut [Rgb565]),
+    ) {
+        fill_buffer(self.buffer);
+
+        for x in &mut self.buffer[dirty_region.min_x() as _..dirty_region.max_x() as _] {
+            *x = embedded_graphics::pixelcolor::raw::RawU16::from(x.into_storage().to_be()).into()
+        }
+        let (ch, mut b, spi) = self.pio.take().unwrap().wait();
+        self.stolen_pin.1.set_high().unwrap();
+
+        /*self.display.set_pixels(
+            dirty_region.min_x() as _,
+            line.get() as _,
+            dirty_region.max_x() as u16,
+            line.get() as u16,
+            self.buffer[dirty_region.origin.x as usize
+                ..dirty_region.origin.x as usize + dirty_region.size.width as usize]
+                .iter()
+                .map(|x| embedded_graphics::pixelcolor::raw::RawU16::from(*x).into_inner()),
+        );*/
+
+        core::mem::swap(&mut self.buffer, &mut b);
```

Comment:
Took me a while to understand this swap. Pretty cool that you can continue rendering the next line during the transfer.